src/musig.rs

```diff
@@ -0,0 +1,1286 @@
+    /// # let sk2 = SecretKey::new(&mut thread_rng());
+    /// # let pub_key2 = PublicKey::from_secret_key(&secp, &sk2);
+    /// #
+    /// let key_agg_cache = MusigKeyAggCache::new(&secp, &[pub_key1, pub_key2]);
+    /// // The session id must be sampled at random. Read documentation for more details.
+    /// let session_id = MusigSessionId::new(&mut thread_rng());
+    ///
+    /// let msg = Message::from_digest_slice(b"Public Message we want to sign!!").unwrap();
+    ///
+    /// // Provide the current time for mis-use resistance
+    /// let extra_rand : Option<[u8; 32]> = None;
+    /// let (_sec_nonce, _pub_nonce) = key_agg_cache.nonce_gen(&secp, session_id, pub_key1, msg, extra_rand)
+    ///     .expect("non zero session id");
+    /// # }
+    /// ```
+    pub fn nonce_gen<C: Signing>(
+        &self,
+        secp: &Secp256k1<C>,
+        session_id: MusigSessionId,
+        pub_key: PublicKey,
+        msg: Message,
+        extra_rand: Option<[u8; 32]>,
+    ) -> Result<(MusigSecNonce, MusigPubNonce), MusigNonceGenError> {
+        // The secret key here is supplied as NULL. This is okay because we supply the
+        // public key and the message.
+        // This makes a simple API for the user because it does not require them to pass here.
+        new_musig_nonce_pair(
+            secp,
+            session_id,
+            Some(&self),
+            None,
+            pub_key,
+            Some(msg),
+            extra_rand,
+        )
+    }
+
+    /// Get a const pointer to the inner MusigKeyAggCache
+    pub fn as_ptr(&self) -> *const ffi::MusigKeyAggCache {
+        &self.0
+    }
+
+    /// Get a mut pointer to the inner MusigKeyAggCache
+    pub fn as_mut_ptr(&mut self) -> *mut ffi::MusigKeyAggCache {
+        &mut self.0
+    }
+}
+
+/// Musig Secret Nonce.
+///
+/// This structure MUST NOT be copied or
+/// read or written to it directly. A signer who is online throughout the whole
+/// process and can keep this structure in memory can use the provided API
+/// functions for a safe standard workflow. See
+/// <https://blockstream.com/2019/02/18/musig-a-new-multisignature-standard/> for
+/// more details about the risks associated with serializing or deserializing
+/// this structure. There are no serialization and parsing functions (yet).
+///
+/// Note this deliberately does not implement `Copy` or `Clone`. After creation, the only
+/// use of this nonce is [`MusigSession::partial_sign`] API that takes ownership of this
+/// and drops it. This is to prevent accidental misuse of this nonce.
+///
+/// A signer who is online throughout the whole process and can keep this
+/// structure in memory can use the provided API functions for a safe standard
+/// workflow.
+///
+/// Signers that pre-computes and saves these nonces are not yet supported. Users
+/// who want to serialize this must use unsafe rust to do so.
+#[allow(missing_copy_implementations)]
+#[derive(Debug, Eq, PartialEq)]
```

Comment:
Shouldn't be derived, we should only display hash or something.

Reply:
Done in https://github.com/rust-bitcoin/rust-secp256k1/pull/716/commits/379f70a152c5be546c678b4179b2ef99c7d955fb